python/pyiceberg/avro/encoder.py

```diff
@@ -0,0 +1,175 @@
+        """Decimal in fixed are encoded as size of fixed bytes."""
+        sign, digits, _ = datum.as_tuple()
+
+        unscaled_datum = 0
+        for digit in digits:
+            unscaled_datum = (unscaled_datum * 10) + digit
+
+        bits_req = unscaled_datum.bit_length() + 1
+        size_in_bits = size * 8
+        offset_bits = size_in_bits - bits_req
+
+        mask = 2**size_in_bits - 1
+        bit = 1
+        for _ in range(bits_req):
+            mask ^= bit
+            bit <<= 1
+
+        if bits_req < 8:
+            bytes_req = 1
+        else:
+            bytes_req = bits_req // 8
+            if bits_req % 8 != 0:
+                bytes_req += 1
+        if sign:
+            unscaled_datum = (1 << bits_req) - unscaled_datum
+            unscaled_datum = mask | unscaled_datum
+            for index in range(size - 1, -1, -1):
+                bits_to_write = unscaled_datum >> (8 * index)
+                self.write(bytearray([bits_to_write & 0xFF]))
+        else:
+            for _ in range(offset_bits // 8):
+                self.write(b"\x00")
+            for index in range(bytes_req - 1, -1, -1):
+                bits_to_write = unscaled_datum >> (8 * index)
+                self.write(bytearray([bits_to_write & 0xFF]))
+
+    def write_bytes(self, b: bytes) -> None:
+        """Bytes are encoded as a long followed by that many bytes of data."""
+        self.write_int(len(b))
+        self.write(struct.pack(f"{len(b)}s", b))
+
+    def write_bytes_fixed(self, b: bytes) -> None:
+        """Writes fixed number of bytes."""
+        self.write(struct.pack(f"{len(b)}s", b))
+
+    def write_utf8(self, s: str) -> None:
+        """A string is encoded as a long followed by that many bytes of UTF-8 encoded character data."""
+        self.write_bytes(s.encode("utf-8"))
+
+    def write_date_int(self, d: date) -> None:
+        """
+        Encode python date object as int.
+
+        It stores the number of days from the unix epoch, 1 January 1970 (ISO calendar).
+        """
+        self.write_int(date_to_days(d))
+
+    def write_time_millis_int(self, dt: time) -> None:
+        """
+        Encode python time object as int.
+
+        It stores the number of milliseconds from midnight, 00:00:00.000
+        """
+        self.write_int(int(time_object_to_micros(dt) / 1000))
+
+    def write_time_micros_long(self, dt: time) -> None:
+        """
+        Encode python time object as long.
+
+        It stores the number of microseconds from midnight, 00:00:00.000000
+        """
+        self.write_int(time_object_to_micros(dt))
+
+    def write_timestamp_millis_long(self, dt: datetime) -> None:
+        """
+        Encode python datetime object as long.
+
+        It stores the number of milliseconds from midnight of unix epoch, 1 January 1970.
+        """
+        self.write_int(int(datetime_to_micros(dt) / 1000))
+
+    def write_timestamp_micros_long(self, dt: datetime) -> None:
```

Comment:
There is no check here for datetime's zone. I think we need to validate that there is no zone. We may also need a `write_timestamptz` method. And can we rename this to `write_timestamp`?

Reply:
I've removed these methods. The encoder should only accept the physical types. I've added the check to the write tree.